test/e2e/storage/csi_mock/csi_workload.go

```diff
@@ -89,36 +88,89 @@ var _ = utils.SIGDescribe("CSI Mock workload info", func() {
 
 				ginkgo.DeferCleanup(m.cleanup)
 
-				withVolume := pvcReference
-				if test.expectEphemeral {
-					withVolume = csiEphemeral
-				}
-				_, _, pod := m.createPod(ctx, withVolume)
-				if pod == nil {
-					return
-				}
-				err = e2epod.WaitForPodNameRunningInNamespace(ctx, m.cs, pod.Name, pod.Namespace)
-				framework.ExpectNoError(err, "Failed to start pod: %v", err)
-
-				// If we expect an ephemeral volume, the feature has to be enabled.
-				// Otherwise need to check if we expect pod info, because the content
-				// of that depends on whether the feature is enabled or not.
-				csiInlineVolumesEnabled := test.expectEphemeral
-				if test.expectPodInfo {
-					ginkgo.By("checking for CSIInlineVolumes feature")
-					csiInlineVolumesEnabled, err = testsuites.CSIInlineVolumesEnabled(ctx, m.cs, f.Timeouts, f.Namespace.Name)
-					framework.ExpectNoError(err, "failed to test for CSIInlineVolumes")
-				}
-
-				ginkgo.By("Deleting the previously created pod")
-				err = e2epod.DeletePodWithWait(ctx, m.cs, pod)
-				framework.ExpectNoError(err, "while deleting")
-
-				ginkgo.By("Checking CSI driver logs")
-				err = checkPodLogs(ctx, m.driver.GetCalls, pod, test.expectPodInfo, test.expectEphemeral, csiInlineVolumesEnabled, false, 1)
-				framework.ExpectNoError(err)
+				verifyPodInfo(ctx, m, test.expectPodInfo, test.expectEphemeral, 1)
+
+			})
+		}
+	})
+
+	ginkgo.Context("CSI PodInfoOnMount Update", func() {
+		var (
+			podInfoTrue  = true
+			podInfoFalse = false
+		)
+		tests := []struct {
+			name              string
+			oldPodInfoOnMount *bool
+			newPodInfoOnMount *bool
+		}{
+			{
+				name:              "should not be passed when update from true to false",
+				oldPodInfoOnMount: &podInfoTrue,
+				newPodInfoOnMount: &podInfoFalse,
+			},
+			{
+				name:              "should not be passed when update from true to nil",
+				oldPodInfoOnMount: &podInfoTrue,
+				newPodInfoOnMount: nil,
+			},
+			{
+				name:              "should be passed when update from false to true",
+				oldPodInfoOnMount: &podInfoFalse,
+				newPodInfoOnMount: &podInfoTrue,
+			},
+			{
+				name:              "should be passed when update from nil to true",
+				oldPodInfoOnMount: nil,
+				newPodInfoOnMount: &podInfoTrue,
+			},
+		}
+		for _, t := range tests {
+			test := t
+			ginkgo.It(t.name, func(ctx context.Context) {
+				m.init(ctx, testParameters{
+					registerDriver: true,
+					podInfo:        test.oldPodInfoOnMount})
+
+				ginkgo.DeferCleanup(m.cleanup)
+
+				verifyPodInfo(ctx, m, test.oldPodInfoOnMount != nil && *test.oldPodInfoOnMount, false, 1)
+				m.update(utils.PatchCSIOptions{PodInfo: test.newPodInfoOnMount})
+				verifyPodInfo(ctx, m, test.newPodInfoOnMount != nil && *test.newPodInfoOnMount, false, 2)
+
 			})
 		}
 	})
 
 })
+
+func verifyPodInfo(ctx context.Context, m *mockDriverSetup, expectPodInfo, expectEphemeral bool, expectedNumNodePublish int) {
```

Comment:
Would this also have the problem where we may need to retry a few times for the change to take effect?

Reply:
Not sure, but the original code doesn't have any retry or sleep mechanism, here I just wrap it in a function.

Reply:
I think this would have the same flaky issue (even today). Would you be able to wrap this check in a retry as well?

Reply:
Sure, updated